oap-server/server-storage-plugin/storage-iotdb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/iotdb/query/IoTDBAggregationQueryDAO.java

```diff
@@ -0,0 +1,110 @@
+import org.apache.skywalking.oap.server.core.storage.query.IAggregationQueryDAO;
+import org.apache.skywalking.oap.server.storage.plugin.iotdb.IoTDBClient;
+
+@Slf4j
+@RequiredArgsConstructor
+public class IoTDBAggregationQueryDAO implements IAggregationQueryDAO {
+    private final IoTDBClient client;
+
+    @Override
+    public List<SelectedRecord> sortMetrics(TopNCondition condition, String valueColumnName, Duration duration,
+                                            List<KeyValue> additionalConditions) throws IOException {
+        // This method maybe have poor efficiency. It queries all data which meets a condition without aggregation function.
+        // https://github.com/apache/iotdb/issues/4006
+        StringBuilder query = new StringBuilder();
+        query.append(String.format("select %s from ", valueColumnName));
+        query = client.addModelPath(query, condition.getName());
+        query.append(" where ").append(IoTDBClient.TIME).append(" >= ").append(duration.getStartTimestamp())
+                .append(" and ").append(IoTDBClient.TIME).append(" <= ").append(duration.getStartTimestamp());
+        if (additionalConditions != null) {
+            for (KeyValue additionalCondition : additionalConditions) {
+                query.append(" and ").append(additionalCondition.getKey()).append(" = \"")
+                        .append(additionalCondition.getValue()).append("\"");
+            }
+        }
+        query.append(IoTDBClient.ALIGN_BY_DEVICE);
+
+        SessionPool sessionPool = client.getSessionPool();
+        SessionDataSetWrapper wrapper = null;
+        List<SelectedRecord> topEntities = new ArrayList<>();
+        try {
+            wrapper = sessionPool.executeQueryStatement(query.toString());
+            if (log.isDebugEnabled()) {
+                log.debug("SQL: {}, columnNames: {}", query, wrapper.getColumnNames());
+            }
+
+            Map<String, Double> entityIdAndSumMap = new HashMap<>();
+            Map<String, Integer> entityIdAndCountMap = new HashMap<>();
+            while (wrapper.hasNext()) {
+                RowRecord rowRecord = wrapper.next();
+                List<Field> fields = rowRecord.getFields();
+                String[] layerNames = fields.get(0).getStringValue().split("\\" + IoTDBClient.DOT + "\"");
+                String entityId = client.layerName2IndexValue(layerNames[2]);
+                double value = Double.parseDouble(fields.get(1).getStringValue());
+                entityIdAndSumMap.merge(entityId, value, Double::sum);
+                entityIdAndCountMap.merge(entityId, 1, Integer::sum);
+            }
+
+            entityIdAndSumMap.forEach((String entityId, Double sum) -> {
+                double count = entityIdAndCountMap.get(entityId);
+                double avg = sum / count;
+                SelectedRecord topNEntity = new SelectedRecord();
+                topNEntity.setId(entityId);
+                topNEntity.setValue(String.valueOf(avg));
```

Comment:
Cast this to `long`

```suggestion
                topNEntity.setValue(String.valueOf((long) avg));
```

Reply:
We only have `long` type value in metrics so this double should be cast to `long`

Reply:
ok, thanks you for pointing out.